### PR TITLE
fix(mac): align native audio startup timestamps

### PR DESCRIPTION
**Pull request by vitaligusatinsky**
## What changed

- Rebase the first microphone and system-audio samples onto the capture timeline when the native macOS source starts 0–250 ms late.
- Keep negative offsets and delays above 250 ms untouched so a failed or intentionally late source is not hidden.
- Add focused Swift tests for zero, negative, bounded, and excessive startup offsets.

## Why

Real native macOS recordings showed audio beginning 159–166 ms after video even though packet cadence was stable. The offset came from per-source startup warm-up, not ongoing drift.

## Impact

New native macOS recordings start audio and video on the same timeline. The correction is deliberately bounded and does not alter spacing between audio packets.

## Validation

- 4 AudioStartAlignment Swift tests pass.
- Before: audio started 159–166 ms after video in two real recordings.
- After: an instrumented recording starts both tracks at 0.000000.
- AAC packet spacing remains exactly 1024/48000 (21.333 ms).

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved audio synchronization at recording startup by compensating for brief initial delays from each audio source.
  * Preserved meaningful timing gaps and delays beyond the startup window.
  * Improved handling of multiple audio sources and late-arriving audio.
  * Added fallback behavior when an enabled source does not provide audio.
  * Improved microphone gain handling to prevent invalid audio levels.

* **Tests**
  * Added automated coverage for startup alignment, source timing, invalid inputs, longer delays, and mixed audio output.

### CODE REVIEW

**Comment by vitaligusatinsky** [comment]
@coderabbitai review

File: electron/native/screencapturekit/Sources/OpenScreenScreenCaptureKitHelper/AudioTrackMixer.swift [hunk 1]
```diff
@@ -110,20 +153,20 @@ final class AudioTrackMixer {
...
+			return
 		}
+
 		guard let anchor else {
 			return
 		}
 
 		let offset = CMTimeConvertScale(
-			CMTimeSubtract(presentationTime, anchor),
+			CMTimeSubtract(alignedPresentationTime, anchor),
```
[the remark this answers is not in the document]
Comment (vitaligusatinsky) [reply electron/native/screencapturekit/Sources/OpenScreenScreenCaptureKitHelper/AudioTrackMixer.swift:196]: Fixed in 027f93b0. The mixer now holds frame-zero emission until every enabled source has delivered its first aligned buffer, with the existing 12,000-frame stall tolerance as the bounded absent-source fallback. Added mixer-level regression coverage proving a late microphone contributes to the first mixed PCM sample at the session start, plus a fallback test. Independent clean Swift run: 6 tests passed.

**Comment by vitaligusatinsky** [comment]
@coderabbitai review

**Comment by vitaligusatinsky** [comment]
Follow-up commit 27a256bc adds the requested startup-boundary coverage: zero offset, negative offset, and the inclusive 250 ms compensation boundary. Clean isolated SwiftPM verification passed all 9 AudioStartAlignment tests, and an independent verifier passed the exact diff.

File: electron/native/screencapturekit/Sources/OpenScreenScreenCaptureKitHelper/AudioTrackMixer.swift [hunk 2]
```diff
@@ -74,7 +116,28 @@ final class AudioTrackMixer {
...
-		self.input = input
+		self.isOutputReady = { input.isReadyForMoreMediaData }
+		self.appendOutput = { input.append($0) }
+		self.includesSystemAudio = includesSystemAudio
+		self.includesMicrophone = includesMicrophone
+		// The request carries MIC_GAIN_BOOST (1.4); Windows applies it unconditionally and so
+		// does this. A non-finite or negative value would poison every mixed sample.
+		let sanitized = microphoneGain.isFinite ? max(0, microphoneGain) : 1
+		self.microphoneGain = Float(sanitized)
+		self.outputFormatDescription = Self.makeOutputFormatDescription()
```
[the remark this answers is not in the document]
Comment (vitaligusatinsky) [reply electron/native/screencapturekit/Sources/OpenScreenScreenCaptureKitHelper/AudioTrackMixer.swift:126]: Addressed in b843c1b9: both initializers now clamp finite values to Float.greatestFiniteMagnitude before conversion through a shared sanitizer. The regression test uses zero PCM first, which traps under the old infinity path, then verifies a loud sample clips safely. Clean and independent SwiftPM runs passed 10/10.

**Comment by vitaligusatinsky** [comment]
Exact-head readiness checkpoint: 98af27024 is rebased onto canonical main fa9719a0 (0 behind / 4 feature commits ahead). Fresh Swift verification passed 10/10, the branch diff is mergeable, and independent implementation review returned GO. Landing remains externally blocked by the upstream maintainer-approval policy and Actions jobs that the repository does not authorize/run for this fork; no merge was attempted. The local task worktree/build cache has been removed because the exact branch is remote-backed.

**Comment by EtienneLescot** [comment]
Closing this one, and I want to be precise about why, because the diagnosis in it is right and the code is careful — the reason is that `main` has since answered the same question a different way, and this branch's answer can't be checked.

## What this gets right, and keeps

Moving the anchor is the good part. At the branch point, `anchor` was set lazily on the first decoded buffer to `CMTimeMaximum(presentationTime, sessionStart)`, so a first audio buffer arriving four seconds late made the mixed track *begin* at four seconds — the leading silence was simply absent from the file. This sets it eagerly to `sessionStart` in `beginTimeline`, and `drain` pads silent chunks from frame zero.

That is exactly the invariant `main`'s new Windows audio-timeline test now asserts in words: *the silence before the tone has to be in the file, not skipped over*. Two people found the same bug on two platforms independently. That part should land.

## Why it isn't landing as written

**The 250 ms window has no support, and the file argues against it.** `stallToleranceFrames = sampleRate / 4` is documented as *"two orders of magnitude above the delivery skew between two outputs of the same SCStream"* — which asserts that skew is around 2.5 ms. `maximumCompensatedStartupDelay` is the same 250 ms, and the tests exercise first-buffer delays of 120 ms and 210 ms. If the 2.5 ms figure is right, those delays are not ordinary SCStream skew, and the compensation window is a hundred times wider than the phenomenon it claims to correct. The branch doesn't reconcile the two numbers, and nothing in the tree does.

**The rebase is per source, so it can desync the two sources against each other.** `AudioStartAlignment` holds a first-PTS per source and derives `sourceOrigin` per source, so each is shifted by its own delay. The test asserts this as intended: system audio's first buffer at 50.12, the microphone's at 50.21, both rebased onto a session start of 50.00 — the 90 ms between them deliberately deleted. That could not happen before: one shared anchor, every source placed by raw PTS against it.

That deletion is sound only if both delays are warm-up. `main` has since established that the two source classes have categorically different delivery semantics — the loopback tap delivers nothing at all while nothing is playing, while a working microphone streams continuously. Under that asymmetry the microphone's first buffer marks when it went live, but the system tap's first buffer marks when a sound first *played*. Treating both as warm-up can move genuine leading audio up to 250 ms early.

**And that is the deeper divergence.** `main` resolved the "was that warm-up, or was it silence?" ambiguity by refusing to infer anything from arrival — the emit counter is driven off `steady_clock` now. This branch resolves the same ambiguity by inferring warm-up from arrival whenever the first buffer lands inside 250 ms. Same question, opposite discipline. That is a decision the repo should make once, on purpose, not arrive at by having each platform patched separately.

**Nothing here can be falsified.** No workflow runs `swift test` — I checked, there is no Swift job in CI at all — so the 301 lines of tests beside `AudioTrackMixer.swift` never execute on a PR. There is no macOS counterpart to `test:wgc-audio-timeline:win`, and that script self-skips off Windows and uses `TOLERANCE_S = 1.0`, so even ported it could not detect a 250 ms mis-rebase. A change to recording timing that no measurement can contradict is not one to merge into a shipped capture path.

**One thing it does not address**, worth recording for whoever picks this up: `drain()` is called only from `ingest` and `finish`, with no clock, and there is no `writer.endSession(atSourceTime:)` anywhere in the helper — so the mixed track still ends at the last delivered sample rather than at the end of the take. If ScreenCaptureKit's system-audio output is silence-gapped the way WASAPI loopback is, a ten-second take that goes quiet at six seconds still yields a six-second audio track. That is the trailing half of the same bug, and `main`'s Windows fix covers it while this does not.

## What would land

The macOS equivalent of what `main` now does on Windows: drive the mixer's emit cursor off a monotonic clock anchored at the writer session start, rather than inferring position from when buffers arrive. That gets the leading silence, the trailing silence, and the mid-take gaps in one mechanism, with no window to justify and no per-source shift to defend. `electron/native/wgc-capture/src/audio_sample_utils.cpp` is the reference implementation, and `scripts/test-windows-audio-timeline.mjs` is the shape of the proof it needs — a macOS sibling of that script would make the whole question decidable.

@vitaligusatinsky — thank you for this, genuinely. Finding that the startup offset was being carried into the file is the hard part, and you found it before we did on the platform we were watching. Closing this is about the mechanism and the absence of a way to verify it, not about the analysis. If you want to come back to it with the clock-driven approach, I will review it properly; and the same offer stands on the audio mastering from #344, which is still a good feature looking for a design that keeps preview and export honest.
